zerver/tests/test_unread.py

```diff
@@ -88,102 +23,12 @@ def test_use_first_unread_anchor_interaction_with_pointer(self) -> None:
-        self.assertFalse(user_message.flags.read)
-
-        # Now if we call get_messages with use_first_unread_anchor=True,
-        # we should get the old message we just set to unread
-        messages_response = self.get_messages_response(
-            anchor="first_unread", num_before=0, num_after=1)
-        self.assertEqual(messages_response['messages'][0]['id'], old_message_id)
-        self.assertEqual(messages_response['anchor'], old_message_id)
-
-        # Let's update the pointer to be *after* this old unread message (but
-        # still on or before the new unread message we just sent)
-        result = self.client_post("/json/users/me/pointer",
-                                  {"pointer": next_old_message_id})
-        self.assert_json_success(result)
-        self.assertEqual(self.example_user('hamlet').pointer,
-                         next_old_message_id)
-
-        # Verify that moving the pointer didn't mark our message as read.
-        user_message = UserMessage.objects.get(
-            message_id=old_message_id,
-            user_profile=self.example_user('hamlet'))
-        self.assertFalse(user_message.flags.read)
-
-    def test_visible_messages_use_first_unread_anchor(self) -> None:
-        self.login('hamlet')
-        self.assertEqual(self.example_user('hamlet').pointer, -1)
-
-        result = self.client_post("/json/mark_all_as_read")
-        self.assert_json_success(result)
-
-        new_message_id = self.send_stream_message(self.example_user("othello"), "Verona",
-                                                  "test")
-
-        messages_response = self.get_messages_response(
-            anchor="first_unread", num_before=0, num_after=1)
-        self.assertEqual(messages_response['messages'][0]['id'], new_message_id)
-        self.assertEqual(messages_response['anchor'], new_message_id)
-
-        with mock.patch('zerver.views.message_fetch.get_first_visible_message_id',
-                        return_value=new_message_id):
-            messages_response = self.get_messages_response(
-                anchor="first_unread", num_before=0, num_after=1)
-        self.assertEqual(messages_response['messages'][0]['id'], new_message_id)
-        self.assertEqual(messages_response['anchor'], new_message_id)
-
-        with mock.patch('zerver.views.message_fetch.get_first_visible_message_id',
-                        return_value=new_message_id + 1):
-            messages_reponse = self.get_messages_response(
-                anchor="first_unread", num_before=0, num_after=1)
-        self.assert_length(messages_reponse['messages'], 0)
-        self.assertIn('anchor', messages_reponse)
-
-        with mock.patch('zerver.views.message_fetch.get_first_visible_message_id',
-                        return_value=new_message_id - 1):
-            messages = self.get_messages(
-                anchor="first_unread", num_before=0, num_after=1)
-        self.assert_length(messages, 1)
```

Comment:
I think these last couple tests are actually important; they're the only results-based testing of the `first_visible_message_id`  and `anchor="first_unread"` logic we had.  

(In `test_narrow` we have tests of the database queries generated, but that isn't the same).  

I think we can just keep them as they were and just remove the pointer blocks from the first one and get a better result.

Reply:
I restored both tests and added a comment.  I think the tests are worth a re-visit from you after the release goes out, as they seem to have somewhat overlapping responsibilities (which made sense in the test-with-and-without-pointer-interaction paradigm).  I don't know enough about the intent of the tests to clean them up effectively.  But I do think they're mostly harmless and may catch future regressions.